src/cascadia/TerminalApp/TerminalPage.cpp

```diff
@@ -1181,36 +1181,78 @@ namespace winrt::TerminalApp::implementation
         // _ClearNewTabButtonColor();
     }
 
+    // Method Description:
+    // - Updates the command palette (tab switcher) with a list of actions
+    //   reflecting the current in-order list of tabs.
+    void TerminalPage::_UpdatePaletteWithInOrderTabs()
+    {
+        auto tabCommands = winrt::single_threaded_vector<Command>();
+        for (const auto& tab : _tabs)
+        {
+            tabCommands.Append(tab.SwitchToTabCommand());
+        }
+        CommandPalette().SetInOrderTabActions(tabCommands);
+    }
+
     // Method Description:
     // - Sets focus to the tab to the right or left the currently selected tab.
     void TerminalPage::_SelectNextTab(const bool bMoveRight)
     {
-        if (_settings.GlobalSettings().UseTabSwitcher())
-        {
-            CommandPalette().SetTabActions(_mruTabActions);
+        const auto tabSwitchMode = _settings.GlobalSettings().TabSwitcherMode();
+        const bool useInOrderTabIndex = tabSwitchMode != TabSwitcherMode::MostRecentlyUsed;
 
-            // Since ATS is always MRU, our focused tab index is always 0.
-            // So, going next should go to index 1, and going prev should wrap to the end.
+        // First, determine what the index of the newly selected tab should be.
+        // This changes if we're doing an in-order traversal vs a MRU traversal.
+        auto newTabIndex = 0;
+        if (useInOrderTabIndex)
+        {
+            // Determine what the next in-order tab index is
+            if (auto index{ _GetFocusedTabIndex() })
+            {
+                uint32_t tabCount = _tabs.Size();
+                // Wraparound math. By adding tabCount and then calculating
+                // modulo tabCount, we clamp the values to the range [0,
+                // tabCount) while still supporting moving leftward from 0 to
+                // tabCount - 1.
+                newTabIndex = ((tabCount + *index + (bMoveRight ? 1 : -1)) % tabCount);
+            }
+        }
+        else
+        {
+            // Determine what the next "most recently used" index is.
+            // In this case, our focused tab index (in the MRU ordering) is
+            // always 0. So, going next should go to index 1, and going prev
+            // should wrap to the end.
             uint32_t tabCount = _mruTabActions.Size();
-            auto newTabIndex = ((tabCount + (bMoveRight ? 1 : -1)) % tabCount);
+            newTabIndex = ((tabCount + (bMoveRight ? 1 : -1)) % tabCount);
+        }
+
+        const bool useTabSwitcher = tabSwitchMode != TabSwitcherMode::Disabled;
+
+        if (useTabSwitcher)
+        {
+            // Set up the list of in-order tabs
+            _UpdatePaletteWithInOrderTabs();
+            // Also set up the list of MRU tabs
+            CommandPalette().SetMRUTabActions(_mruTabActions);
```

Comment:
Curious why we don't just give the cmdpal the requested set of actions here instead of both? Is it so the ListView animations for removals/inserts work?

Reply:
Both you and Michael called this one out, and frankly, I think it was mostly just "me prototyping at 5am". _There's gotta be a way_ to do it with just one list of actions.